cpp/src/arrow/ipc/writer.h

```diff
@@ -24,6 +24,7 @@
 #include <memory>
 #include <vector>
 
+#include "arrow/ipc/dictionary.h"
```

Comment:
Could instead forward-declare class `DictionaryMemo` (not sure how important it is, but if we do this often it helps reduce compile times).